File: common/src/main/java/io/xol/chunkstories/workers/TasksPool.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.workers;

import java.util.Deque;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicInteger;

import io.xol.chunkstories.api.workers.Task;

/**
 * A task pool receives tasks and dispatches them
 */
public abstract class TasksPool<T extends Task>
{
	protected Deque<T> tasksQueue = new ConcurrentLinkedDeque<T>();
	protected Semaphore tasksCounter = new Semaphore(0);
	protected AtomicInteger tasksQueueSize = new AtomicInteger(0);
	
	public void scheduleTask(T task)
	{
		//Add the tasks BEFORE doing the thing
		tasksQueue.add(task);
		tasksCounter.release();
		tasksQueueSize.incrementAndGet();
	}
	
	public int size()
	{
		return tasksQueueSize.get();
	}
}

File: common/src/main/java/xyz/chunkstories/task/TasksPool.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//
package xyz.chunkstories.task

import xyz.chunkstories.api.workers.Task
import java.util.Deque
import java.util.concurrent.ConcurrentLinkedDeque
import java.util.concurrent.Semaphore
import java.util.concurrent.atomic.AtomicInteger

abstract class TasksPool<T : Task?> {
    internal var tasksQueue: Deque<T> = ConcurrentLinkedDeque()
    internal var tasksCounter = Semaphore(0)
    internal var tasksQueueSize = AtomicInteger(0)

    fun scheduleTask(task: T) {
        tasksQueue.add(task)
        tasksCounter.release()
        tasksQueueSize.incrementAndGet()
    }

    fun size(): Int {
        return tasksQueueSize.get()
    }
}